tests/texture_draw.rs

```diff
@@ -34,30 +30,81 @@ macro_rules! texture_draw_test {
                 ", $glsl_ty, $glsl_value),
                 None);
 
-            let program = match program {
+            match program {
                 Ok(p) => p,
                 Err(_) => return
-            };
-
-            let texture = glium::texture::$tex_ty::empty(&display, $($dims),+).unwrap();
+            }
+        }
+    );
+}
 
-            texture.as_surface().clear_color(0.0, 0.0, 0.0, 0.0);
-            texture.as_surface().draw(&vb, &ib, &program, &uniform!{ texture: &texture },
+macro_rules! draw_and_validate {
+    ($display: expr, $program: expr, $texture:expr, $vb:expr, $ib:expr, $rust_value:expr) => (
+        {
+            $texture.as_surface().clear_color(0.0, 0.0, 0.0, 0.0);
+            $texture.as_surface().draw(&$vb, &$ib, &$program, &uniform!{ texture: &$texture },
                                      &Default::default()).unwrap();
 
-            display.assert_no_error(None);
+            $display.assert_no_error(None);
 
-            let data: Vec<Vec<(u8, u8, u8, u8)>> = texture.read();
+            let data: Vec<Vec<(u8, u8, u8, u8)>> = $texture.read();
             for row in data.iter() {
                 for pixel in row.iter() {
                     assert_eq!(pixel, &$rust_value);
                 }
             }
 
-            display.assert_no_error(None);
+            $display.assert_no_error(None);
+        }
+    );
+}
+
+macro_rules! texture_draw_test {
+    ($test_name:ident, $tex_ty:ident, [$($dims:expr),+], $glsl_ty:expr, $glsl_value:expr,
+     $rust_value:expr) => (
+        #[test]
+        fn $test_name() {
+            let display = support::build_display();
+            let (vb, ib) = support::build_rectangle_vb_ib(&display);
+
+            let program = create_program!(display, $glsl_ty, $glsl_value);
+
+            let texture = glium::texture::$tex_ty::empty(&display, $($dims),+).unwrap();
+
+            draw_and_validate!(display, program, texture, vb, ib, $rust_value);
+        }
+    );
+}
+
+macro_rules! unowned_draw_test {
+    ($test_name:ident, $tex_ty:ident, $format_ty:ident, $format_value:ident, [$($dims:expr),+], $glsl_ty:expr, $glsl_value:expr,
+     $rust_value:expr) => (
+        #[test]
+        fn $test_name() {
+            let display = support::build_display();
+            let (vb, ib) = support::build_rectangle_vb_ib(&display);
+
+            let program = create_program!(display, $glsl_ty, $glsl_value);
+
+            let empty_texture = glium::texture::$tex_ty::empty_with_format(&display,
+                                                                           glium::texture::$format_ty::$format_value,
+                                                                           glium::texture::MipmapsOption::NoMipmap,
+                                                                           $($dims),+).unwrap();
+            let texture = unsafe {
+                glium::texture::$tex_ty::from_id(&display,
```

Comment:
I assume it's safe to just repeat the format here and trust that Glium didn't change the underlying format, as I can't find an easy way to convert the internal format of a texture back out into an UncompressedFloatFormat.